ui/app/hooks/useRetryTransaction.js

```diff
@@ -0,0 +1,110 @@
+import { useDispatch } from 'react-redux'
+import { useMemo, useCallback, useEffect, useState } from 'react'
+import * as actions from '../store/actions'
+import {
+  fetchBasicGasAndTimeEstimates as fetchBasicGasAndTimeEstimatesAction,
+  fetchGasEstimates as fetchGasEstimatesAction,
+  setCustomGasPriceForRetry as setCustomGasPriceForRetryAction,
+  setCustomGasLimit as setCustomGasLimitAction,
+} from '../ducks/gas/gas.duck'
+import { TOKEN_METHOD_TRANSFER } from '../helpers/constants/transactions'
+import { increaseLastGasPrice } from '../helpers/utils/confirm-tx.util'
+import { useMetricEvent } from './useMetricEvent'
+import { useMethodData } from './useMethodData'
+
+
+/**
+ * useRetryTransaction
+ *
+ * Provides a reusable hook that, given a transactionGroup, will return
+ * whether or not the transaction meets the criteria to be retried with
+ * higher gas (sped up), and a method for beginning the retry process
+ * @param {Object} transactionGroup - the transaction group
+ * @param {boolean} isEarliestNonce - indicates if this transaction is the earliest in history
+ * @return {[boolean, Function]}
+ */
+export function useRetryTransaction (transactionGroup, isEarliestNonce = false) {
+  const { primaryTransaction, transactions, initialTransaction, hasRetried } = transactionGroup
+  const [earliestTransaction = {}] = transactions
+  const { submittedTime } = earliestTransaction
+  const gasPrice = primaryTransaction.txParams?.gasPrice
+  const methodData = useMethodData(primaryTransaction.txParams?.data)
+  const trackMetricsEvent = useMetricEvent(({
+    eventOpts: {
+      category: 'Navigation',
+      action: 'Activity Log',
+      name: 'Clicked "Speed Up"',
+    },
+  }))
+  const dispatch = useDispatch()
+  const { fetchBasicGasAndTimeEstimates, fetchGasEstimates, setSelectedToken, retryTransaction } = useMemo(() => ({
+    fetchBasicGasAndTimeEstimates: () => dispatch(fetchBasicGasAndTimeEstimatesAction()),
+    fetchGasEstimates: (blockTime) => dispatch(fetchGasEstimatesAction(blockTime)),
+    setSelectedToken: (tokenAddress) => dispatch(actions.setSelectedToken(tokenAddress)),
+    retryTransaction: () => {
+      const transaction = initialTransaction
+      const increasedGasPrice = increaseLastGasPrice(gasPrice)
+      dispatch(setCustomGasPriceForRetryAction(increasedGasPrice || transaction.txParams?.gasPrice))
```

Comment:
in this specific case or anywhere? e2e failed pretty heavily with errors about accessing a value on undefined with txParams. I'll remove this and see if i can get you better details.

Reply:
I _think_ `txParams` should always exist on actual transactions? It'd be missing from signature requests, but I don't see how this hook would get used for those.

Reply:
@Gudahtt its probably signing requests because hooks are all called at top level and some work is done inside the top level of this hook:

```
export function useRetryTransaction (transactionGroup) {
  const { primaryTransaction, initialTransaction } = transactionGroup
  const gasPrice = primaryTransaction.txParams?.gasPrice
```
i probably then got overzealous by adding ? marks anywhere I saw txParams. it should always be defined inside this callback. I'll see if i can remove this one.

Reply:
Ah right. Ok, that would make sense. So this would still get run for signature request items, but the callback would never be used.

Keeping the `?.` is totally fine with me - I was just curious.